### DISCUSSION TITLE
Multi-tenant with hooks for all operations?

### DISCUSSION THREAD

**Discussion by frederikhors** [ent/ent#2585] [Q&A] [answered]
I'm coming from [Bun](https://github.com/uptrace/bun) which has Hooks too:

It is very easy with Bun to add a tenanted hook on each operation (even the select one):

```go
type Player struct {
  Id   uint64
  Name string
}

var _ bun.BeforeSelectHook = (*Player)(nil)

func (*Player) BeforeSelect(ctx context.Context, query *bun.SelectQuery) error {
	query.Where("?TableAlias.tenant_id = ?", GetTenantID(ctx))

	return nil
}

var _ bun.BeforeInsertHook = (*Player)(nil)

func (*Player) BeforeInsert(ctx context.Context, query *bun.InsertQuery) error {
	query.Value("?TableAlias.tenant_id = ?", GetTenantID(ctx))

	return nil
}

var _ bun.BeforeUpdateHook = (*Player)(nil)

func (*Player) BeforeUpdate(ctx context.Context, query *bun.UpdateQuery) error {
	query.Where("?TableAlias.tenant_id = ?", GetTenantID(ctx))

	return nil
}

var _ bun.BeforeDeleteHook = (*Player)(nil)

func (*Player) BeforeDelete(ctx context.Context, query *bun.DeleteQuery) error {
	query.Where("?TableAlias.tenant_id = ?", GetTenantID(ctx))

	return nil
}
```

In this case I can rest assured that no one on the team will forget to add a `query.Where("tenant_id=?")` in the code because the hook saves me in all cases.

I'm learning Ent from today and saw this description on [Readme for Tenant example](https://github.com/ent/ent/tree/master/examples/privacytenant) which is not something I expect for it:

![image](https://user-images.githubusercontent.com/41120635/171388349-a901e017-3011-418e-a032-d5dd2c5cc22a.png)

Is it a misprint?

Can you direct me to a guide for a multi-tenant with hooks for all operations?

I'm a bit scared because in this page https://entgo.io/docs/hooks there are no hook for operations like "Select".

**Comment by a8m** [comment]
Hey @frederikhors, for this use case, you can use the [Privacy Filter Rules](https://entgo.io/docs/privacy/#filter-rules) to apply filtering to all your queries/mutations (including `Select`).

> Is it a misprint?

Fixed. Thanks for reporting this.

**Comment by a8m** [comment, marked as answer]
I implemented with Ent a few different approaches for multi-tenacy. Named-database per tenant, tenant edge, and tenant_id field. The field/edge-based approaches were implemented with FilterRule as follows:

```go
func FilterTenantRule() privacy.QueryRule {
	type TenantFilter interface {
		WhereTenantID(int)
	}
	return privacy.FilterFunc(func(ctx context.Context, f privacy.Filter) error {
		t := viewer.TenantFromContext(ctx)
		tf, ok := f.(FilterTenantRule)
		if !ok {
			return privacy.Denyf("unexpected filter type %T. All schemas must include the Tenant mixin", f)
		}
		tf.WhereTenantID(t.ID)
		// Skip to the next privacy rule (equivalent to return nil).
		return privacy.Skip
	})
}
```

Let me know if it helps or if you need help with it.

**Reply by frederikhors** [reply]
@a8m I managed to get everything to work with the filter rule.

What I can't figure out now is if there is a way to avoid manually using the `.SetTenant()` in the `.Create()` methods.

Can I use a rule for this too?

**Reply by frederikhors** [reply]
I'm using Hooks. Do you think is the way to go?

**Reply by frederikhors** [reply]
For future people: [I summarized issues here](https://github.com/ent/ent/issues/2601).
